inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/heartbeat/HeartbeatManager.java

```diff
@@ -162,6 +177,23 @@ public void reportHeartbeat(HeartbeatMsg heartbeat) {
                     handlerNum += insertClusterNode(clusterInfo, heartbeatMsg, clusterInfo.getCreator());
                 } else {
                     handlerNum += updateClusterNode(clusterNode, heartbeatMsg);
+                    // If the agent report succeeds, restore the source status
+                    if (Objects.equals(clusterNode.getType(), ClusterType.AGENT)) {
+                        List<StreamSourceEntity> sourceEntities = sourceMapper.selectByAgentIpAndCluster(
+                                Collections.singletonList(SourceStatus.HEARTBEAT_TIMEOUT.getCode()),
+                                Lists.newArrayList(SourceType.FILE), heartbeat.getIp(), heartbeatMsg.getClusterName());
+                        for (StreamSourceEntity sourceEntity : sourceEntities) {
+                            // restore state for all source by ip and type
+                            if (sourceEntity.getIsDeleted() != 0) {
+                                sourceEntity.setPreviousStatus(sourceEntity.getStatus());
+                                sourceEntity.setStatus(SourceStatus.TO_BE_ISSUED_DELETE.getCode());
+                            } else {
+                                sourceEntity.setStatus(sourceEntity.getPreviousStatus());
+                                sourceEntity.setPreviousStatus(SourceStatus.HEARTBEAT_TIMEOUT.getCode());
+                            }
+                            sourceMapper.updateByPrimaryKeySelective(sourceEntity);
```

Comment:
It is a bad practice to read and write DB in a loop.
Can you change to update records in batches?